pyk/src/pyk/kcfg/kcfg.py

```diff
@@ -569,6 +570,20 @@ def log(message: str, *, warning: bool = False) -> None:
                 f'Extending current KCFG with the following: {message}{result_info_message}',
             )
 
+        def optimize_step(
+            cterm: CTerm, depth: int, next_node_logs: tuple[LogEntry, ...], rule_labels: list[str]
+        ) -> bool:
+            in_edges = self.edges(target_id=node.id)
+            if len(in_edges) == 1:
+                in_edge = single(in_edges)
+                self.remove_edge(in_edge.source.id, node.id)
```

Comment:
Does this code have to live here or could it be in `APRProver.step_proof` or `APRProver.commit` instead? Passing the data in the `APRProofResult` is better, it keeps our existing abstraction of only passing `*Step` and `*Result` objects back and forth between master and worker threads.

If it lives in `APRProver`, then it can access `KCFGMinimizer` (defined in `minimizepy`) and re-use the routines there instead of making a new one.

Reply:
For example, it can reuse this routine: https://github.com/runtimeverification/k/blob/ca93ca476a73abfbff358a6911fcb0e260371d58/pyk/src/pyk/kcfg/minimize.py#L53

Reply:
Yes, I am aware of this routine. This is not the same routine, but rather an analogous one. The key difference is that `lift_edge`/`lift_edges` require both/all edges to already exist. Here, we are "extending" a single edge by modifying its `target` in place.

We could create both edges first, but this would mean that the node counter for the KCFG will be incremented, which could result in a KCFG that has edges like `3 -> 200 ->> {201, 202}` instead of `3 -> 4 ->> {5, 6}`. By doing the extension directly, we also keep the node counter at a minimum.

I'm not sure how to move this code to the places you suggested, since the modifications to the KCFG happen in `kcfg.py`.

Reply:
I do recognise the issue of code duplication and will think how to improve.

Reply:
I'm not sure. 

This routine updates one edge (by removing and re-creating it) and updates its `target` node.
The `lift_edge` routine in `minimize.py` removes the central node (effectively removing two edges) and then creates an edge between the outer nodes.

There is little actual duplication, it's just a similar approach.

Reply:
Sonuds good.